tooling/autodispose-rx-error-prone-checker/src/main/java/com/uber/autodispose/rx/error/prone/checker/RxJavaMissingAutoDisposeErrorChecker.java

```diff
@@ -0,0 +1,197 @@
+        .add(instanceMethod().onDescendantOf("io.reactivex.Completable").named(SUBSCRIBE))
+        .add(instanceMethod().onDescendantOf("io.reactivex.Flowable").named(SUBSCRIBE))
+        .add(instanceMethod().onDescendantOf("io.reactivex.Maybe").named(SUBSCRIBE));
+    SUBSCRIBE_MATCHERS = builder3.build();
+  }
+
+  private static final Matcher<ExpressionTree> METHOD_NAME_MATCHERS =
+      new Matcher<ExpressionTree>() {
+        @Override
+        public boolean matches(ExpressionTree tree, VisitorState state) {
+          if (!(tree instanceof MethodInvocationTree)) {
+            return false;
+          }
+          MethodInvocationTree invTree = (MethodInvocationTree) tree;
+
+          final MemberSelectTree memberTree = (MemberSelectTree) invTree.getMethodSelect();
+          if (!memberTree.getIdentifier().contentEquals(TO)) {
+            return false;
+          }
+
+          for (MethodMatchers.MethodNameMatcher nameMatcher : TO_CALL_MATCHERS) {
+            if (nameMatcher.matches(invTree, state)) {
+              ExpressionTree arg = invTree.getArguments().get(0);
+              final Type scoper = state.getTypeFromString("com.uber.autodispose.Scoper");
+              return ASTHelpers.isSubtype(ASTHelpers.getType(arg), scoper, state);
+            }
+          }
+
+          for (MethodMatchers.MethodNameMatcher nameMatcher : AS_CALL_MATCHERS) {
+            if (nameMatcher.matches(invTree, state)) {
+              ExpressionTree arg = invTree.getArguments().get(0);
+              final Type scoper = state
+                  .getTypeFromString("com.uber.autodispose.AutoDisposeConverter");
+              return ASTHelpers.isSubtype(ASTHelpers.getType(arg), scoper, state);
+            }
+          }
+          return false;
+        }
+      };
+
+  private static Matcher<MethodInvocationTree> matcher(List<String> classesWithLifecycle) {
+    return new Matcher<MethodInvocationTree>() {
+      @Override
+      public boolean matches(MethodInvocationTree tree, VisitorState state) {
+
+        boolean matchFound = false;
+        try {
+          final MemberSelectTree memberTree = (MemberSelectTree) tree.getMethodSelect();
+          if (!memberTree.getIdentifier().contentEquals(SUBSCRIBE)) {
+            return false;
+          }
+
+          for (MethodMatchers.MethodNameMatcher nameMatcher : SUBSCRIBE_MATCHERS) {
+            if (!nameMatcher.matches(tree, state)) {
+              continue;
+            } else {
+              matchFound = true;
+              break;
+            }
+          }
+          if (!matchFound) {
+            return false;
+          }
+
+          ClassTree enclosingClass =
+              ASTHelpers.findEnclosingNode(state.getPath(), ClassTree.class);
+          Type.ClassType enclosingClassType = ASTHelpers.getType(enclosingClass);
+
+          for (String s : classesWithLifecycle) {
+            Type lifecycleType = state.getTypeFromString(s);
```

Comment:
Small optimization: store `classesWithLifecycle` as a `List<Type>`.  The call `getTypeFromString()` can be a bit costly (though for this checker it may not get called very often)

Reply:
Missed it. Will fix in a separate commit.

Reply:
@msridhar quick question - do you mean to cache it when we will call it the first time? Can we get the type from a string without visitor state? I am not sure how to do it eagerly.

Reply:
Yeah you're right, hard to do eagerly.  We can skip this optimization for now I think; this code shouldn't run all that often.